controllers/nugus_controller/nugus_controller.cpp

```diff
@@ -15,106 +15,322 @@
+                                    break;
+                                }
+                                case webots::Node::CAMERA: {
+                                    std::unique_ptr<webots::Camera> camera(this->getCamera(sensorTimeStep.name()));
+                                    camera->enable(sensor_time_step);
+                                    break;
+                                }
+                                case webots::Node::GYRO: {
+                                    std::unique_ptr<webots::Gyro> gyro(this->getGyro(sensorTimeStep.name()));
+                                    gyro->enable(sensor_time_step);
+                                    break;
+                                }
+                                case webots::Node::POSITION_SENSOR: {
+                                    std::unique_ptr<webots::PositionSensor> positionSensor(
+                                        this->getPositionSensor(sensorTimeStep.name()));
+                                    positionSensor->enable(sensor_time_step);
+                                    break;
+                                }
+                                case webots::Node::TOUCH_SENSOR: {
+                                    std::unique_ptr<webots::TouchSensor> touchSensor(
+                                        this->getTouchSensor(sensorTimeStep.name()));
+                                    touchSensor->enable(sensor_time_step);
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
 
-                Nh = msg.ByteSizeLong();
-                data.resize(Nh);
-                msg.SerializeToArray(data.data(), Nh);
+    void sendData(int controller_time) {
+        // Create the SensorMeasurements message
+        auto sensorMeasurements = std::make_unique<SensorMeasurements>();
 
-                Nn = htonl(Nh);
+        sensorMeasurements->set_time(controller_time);
+        struct timeval tp;
+        gettimeofday(&tp, NULL);
+        uint64_t real_time = tp.tv_sec * 1000 + tp.tv_usec / 1000;
+        sensorMeasurements->set_real_time(real_time);
 
-                if (send(tcp_fd, &Nn, sizeof(Nn), 0) < 0) {
-                    std::cerr << "Error: Failed to send data over TCP connection: " << strerror(errno) << std::endl;
+        // Iterator over all the devices that have been enabled from any received ActuatorRequests messages
+        for (std::set<std::shared_ptr<webots::Device>>::iterator it = sensors.begin(); it != sensors.end(); ++it) {
+            // Try to cast to an accelerometer and if it works, add the accelerometer values to the SensorMeasurement
+            // message
+            std::shared_ptr<webots::Accelerometer> accelerometer =
+                std::dynamic_pointer_cast<webots::Accelerometer>(*it);
```

Comment:
This is not very likely to fail. All devices (accelerometers, etc) inherit from `webots::Device` and so casting it from a base class to a subclass will probably work.

Using [wb_get_node_type](https://cyberbotics.com/doc/reference/device?tab-language=c++#wb_device_get_node_type) on the device may be the better option